<a href="https://colab.research.google.com/github/Gugunner/pandas-challenge/blob/master/HeroesOfPymoli.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Heroes of Pymoli Homework

![Fantasy](http://drive.google.com/uc?export=view&id=1xqJfwaaZJ_warRjU8BX9PVFoz1aXIm5t)

Congratulations! After a lot of hard work in the data munging mines, you've landed a job as Lead Analyst for an independent gaming company. You've been assigned the task of analyzing the data for their most recent fantasy game Heroes of Pymoli.

Like many others in its genre, the game is free-to-play, but players are encouraged to purchase optional items that enhance their playing experience. As a first task, the company would like you to generate a report that breaks down the game's purchasing data into meaningful insights.

## How to read this homework
You will find that each task has it's instruction before the code to solve it, at the end and before a new task the solution can be found with a visual data frame representation.

At the end of this note book you'll find the observable trends.

In [49]:
# Dependencies imports
import pandas as pd
from google.colab import drive 
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [8]:
# Checking csv for successfull read
csv_path = '/content/gdrive/My Drive/Bootcamp_Data_Analytics/pandas-challenge/HeroesOfPymoli/Resources/purchase_data.csv'
purchase_data = pd.read_csv(csv_path)
purchase_data.head()

Purchase ID             SN  ...                                  Item Name Price
0            0        Lisim78  ...  Extraction, Quickblade Of Trembling Hands  3.53
1            1    Lisovynya38  ...                          Frenzied Scimitar  1.56
2            2     Ithergue48  ...                               Final Critic  4.88
3            3  Chamassasya86  ...                                Blindscythe  3.27
4            4      Iskosia90  ...                                       Fury  1.44

[5 rows x 7 columns]

### Player Count

* Total Number of Players
---------------------------------

In [9]:
# Number of players
number_of_players = len(purchase_data['SN'].value_counts())
number_of_players
players_df = pd.DataFrame({
  'Total Players': [number_of_players]  
})
players_df

Total Players
0            576

### Result Purchasing Analysis (Total)

* Number of Unique Items
* Average Purchase Price
* Total Number of Purchases
* Total Revenue
-----------------------------

In [10]:
# Number of unique items
number_of_unique_items = len(purchase_data['Item Name'].unique())
number_of_unique_items

179

In [11]:
# Average purchase price
avg_price = round(purchase_data['Price'].mean(),2)
avg_price

3.05

In [12]:
# Total number of purchases
total_purchases = purchase_data['Purchase ID'].count()
total_purchases

780

In [13]:
# Total revenue 
total_revenue = purchase_data['Price'].sum()
total_revenue

2379.77

In [14]:
pruchase_analysis_total_df = pd.DataFrame({
    'Number of Unique Items': [number_of_unique_items],
    'Average Price': [avg_price],
    'Number of Purchases': [total_purchases],
    'Total Revenue': [total_revenue]
})
pruchase_analysis_total_df['Average Price'] = pruchase_analysis_total_df['Average Price'].map('${:.2f}'.format)
pruchase_analysis_total_df['Total Revenue'] = pruchase_analysis_total_df['Total Revenue'].map('${:.2f}'.format)
pruchase_analysis_total_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780      $2379.77

### Result Gender Demographics

* Percentage and Count of Male Players
* Percentage and Count of Female Players
* Percentage and Count of Other / Non-Disclosed
-----------

In [15]:
# Percentage and Count of Male Players
male_players = len(purchase_data.loc[purchase_data['Gender'] == 'Male']['SN'].value_counts())
male_players

484

In [16]:
male_pct = round((male_players/number_of_players)*100, 2)
male_pct

84.03

In [17]:
# Percentage and Count of Female Players
female_players = len(purchase_data.loc[purchase_data['Gender'] == 'Female']['SN'].value_counts())
female_players

81

In [18]:
female_pct = round((female_players/number_of_players)*100, 2)
female_pct

14.06

In [19]:
# Percentage and Count of Other/Non-Disclosed
other_players = len(purchase_data.loc[purchase_data['Gender'] == 'Other / Non-Disclosed']['SN'].value_counts())
other_players

11

In [20]:
other_pct = round((other_players/number_of_players)*100, 2)
other_pct

1.91

In [21]:
gender_demo_df = pd.DataFrame([
    {
        'Gender': 'Male',
        'Total Count': male_players,
        'Percentage of Players': male_pct
    },
    {
        'Gender': 'Feale',
        'Total Count': female_players,
        'Percentage of Players': female_pct
    },
    {
        'Gender': 'Others/Non-Disclosed',
        'Total Count': other_players,
        'Percentage of Players': other_pct
    },
])
gender_demo_df = gender_demo_df.set_index('Gender')
gender_demo_df['Percentage of Players'] = gender_demo_df['Percentage of Players'].map('{:.2f}%'.format)
gender_demo_df.index.name = None
gender_demo_df

Total Count Percentage of Players
Male                          484                84.03%
Feale                          81                14.06%
Others/Non-Disclosed           11                 1.91%

### Result Purchasing Analysis (Gender)

* The below each broken by gender
  * Purchase Count
  * Average Purchase Price
  * Total Purchase Value
  * Average Purchase Total per Person by Gender
-------------------

In [22]:
# New male purchase data frame
male_purchase_data = purchase_data.loc[purchase_data['Gender'] == 'Male']
male_purchase_data.head()

Purchase ID             SN  ...                                  Item Name Price
0            0        Lisim78  ...  Extraction, Quickblade Of Trembling Hands  3.53
1            1    Lisovynya38  ...                          Frenzied Scimitar  1.56
2            2     Ithergue48  ...                               Final Critic  4.88
3            3  Chamassasya86  ...                                Blindscythe  3.27
4            4      Iskosia90  ...                                       Fury  1.44

[5 rows x 7 columns]

In [23]:
# Male purchase count
male_purchase = len(male_purchase_data)
male_purchase

652

In [24]:
# Average male purchase price
male_mean_price = round(male_purchase_data['Price'].mean(),2)
male_mean_price

3.02

In [25]:
# Total male purchase 
male_total_purchase = round(male_purchase_data['Price'].sum(),2)
male_total_purchase

1967.64

In [26]:
# Average purchase total per person by males
male_avg_purchase = round(male_total_purchase/male_players,2)
male_avg_purchase

4.07

In [27]:
# New female purchase data frame
female_purchase_data = purchase_data.loc[purchase_data['Gender'] == 'Female']
female_purchase_data.head()

Purchase ID         SN  Age  ... Item ID                     Item Name Price
15           15  Lisassa64   21  ...      98   Deadline, Voice Of Subtlety  2.89
18           18  Reunasu60   22  ...      82                       Nirvana  4.90
38           38   Reulae52   10  ...     116       Renewed Skeletal Katana  4.18
41           41  Assosia88   20  ...       7  Thorn, Satchel of Dark Souls  1.33
55           55  Phaelap26   25  ...      84                    Arcane Gem  3.79

[5 rows x 7 columns]

In [28]:
# Female purchase count
female_purchase = len(female_purchase_data)
female_purchase

113

In [29]:
# Average female purchase price
female_mean_price = round(female_purchase_data['Price'].mean(),2)
female_mean_price

3.2

In [30]:
# Total female purchase 
female_total_purchase = round(female_purchase_data['Price'].sum(),2)
female_total_purchase

361.94

In [31]:
# Average purchase total per person by females
female_avg_purchase = round(female_total_purchase/female_players,2)
female_avg_purchase

4.47

In [32]:
# New other/non-disclosed data frame
other_purchase_data = purchase_data.loc[purchase_data['Gender'] == 'Other / Non-Disclosed']
other_purchase_data.head()

Purchase ID            SN  ...                     Item Name Price
9              9   Chanosian48  ...  Ghastly Adamantite Protector  3.58
22            22  Siarithria38  ...                 Warped Fetish  3.81
82            82    Haerithp41  ...                    Azurewrath  4.40
111          111      Sundim98  ...                         Orbit  4.75
228          228   Jiskirran77  ...                     Dreamsong  3.39

[5 rows x 7 columns]

In [33]:
# Other purchase count
other_purchase = len(other_purchase_data)
other_purchase

15

In [34]:
# Average other purchase price
other_mean_price = round(other_purchase_data['Price'].mean(),2)
other_mean_price

3.35

In [35]:
# Total other purchase
other_total_purchase = round(other_purchase_data['Price'].sum(),2)
other_total_purchase

50.19

In [36]:
# Average purchase total per person by others
other_avg_purchase = round(other_total_purchase/other_players,2)
other_avg_purchase

4.56

In [37]:
purchase_analysis_gender_df = pd.DataFrame({
    'Gender': ['Female', 'Male', 'Other/Non-Disclosed'],
    'Purchase Count': [female_purchase, male_purchase, other_purchase],
    'Average Purchase Price': [female_mean_price, male_mean_price, other_mean_price],
    'Total Purchase Value': [female_total_purchase, male_total_purchase, other_total_purchase],
    'Avg Total Purchase per Person': [female_avg_purchase, male_avg_purchase, other_avg_purchase]
})
purchase_analysis_gender_df['Average Purchase Price'] = purchase_analysis_gender_df['Average Purchase Price'].map('${:.2f}'.format)
purchase_analysis_gender_df['Total Purchase Value'] = purchase_analysis_gender_df['Total Purchase Value'].map('${:.2f}'.format)
purchase_analysis_gender_df['Avg Total Purchase per Person'] = purchase_analysis_gender_df['Avg Total Purchase per Person'].map('${:.2f}'.format)
purchase_analysis_gender_df = purchase_analysis_gender_df.set_index('Gender')
purchase_analysis_gender_df.index.name = None
purchase_analysis_gender_df

Purchase Count  ... Avg Total Purchase per Person
Female                          113  ...                         $4.47
Male                            652  ...                         $4.07
Other/Non-Disclosed              15  ...                         $4.56

[3 rows x 4 columns]

### Result Age Demographics
* Establish bins for ages
* Categorize the existing players using the age bins. Hint: use pd.cut()
* Calculate the numbers and percentages by age group
* Create a summary data frame to hold the results
* Optional: round the percentage column to two decimal points
* Display Age Demographics Table
-----------

In [38]:
# Get min and max age value
purchase_data['Age'].max()

45

In [39]:
purchase_data['Age'].min()

7

In [40]:
age_bins = [0,9, 14, 19, 24, 29, 34, 39, 45]
age_labels = ['<10', '10-14', '15-19', '20-24', '25-29', '30-34', '35-39', '40+']
pd.cut(purchase_data['Age'], age_bins, include_lowest=False, duplicates='drop')

0      (19, 24]
1      (39, 45]
2      (19, 24]
3      (19, 24]
4      (19, 24]
         ...   
775    (19, 24]
776    (19, 24]
777    (19, 24]
778      (0, 9]
779    (19, 24]
Name: Age, Length: 780, dtype: category
Categories (8, interval[int64]): [(0, 9] < (9, 14] < (14, 19] < (19, 24] < (24, 29] < (29, 34] <
                                  (34, 39] < (39, 45]]

In [0]:
new_purchase_data = purchase_data
new_purchase_data['Age Range'] = pd.cut(purchase_data['Age'], age_bins, labels=age_labels, include_lowest=False, duplicates='drop')

In [42]:
total_grouped_df = new_purchase_data.groupby(['Age Range'])
total_grouped_df = total_grouped_df.agg({ 'SN': 'nunique' })
total_grouped_df = total_grouped_df.reset_index()

age_demo_df = pd.DataFrame(total_grouped_df)
age_demo_df['Percentage of Players'] = age_demo_df.apply(lambda row: (row[1]/number_of_players)*100, axis=1)
age_demo_df['Percentage of Players'] = age_demo_df['Percentage of Players'].map('{:.2f}%'.format)
age_demo_df = age_demo_df.rename(columns={
    'SN': 'Total Count'
})
age_demo_df = age_demo_df.set_index('Age Range')
age_demo_df.index.name = None
age_demo_df

Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

### Result Purchasing Analysis (Age)

* The below each broken into bins of 4 years (i.e. &lt;10, 10-14, 15-19, etc.)
  * Purchase Count
  * Average Purchase Price
  * Total Purchase Value
  * Average Purchase Total per Person by Age Group
----------

In [0]:
age_bins = [0,9, 14, 19, 24, 29, 34, 39, 45]
age_labels = ['<10', '10-14', '15-19', '20-24', '25-29', '30-34', '35-39', '40+']
pd.cut(purchase_data['Age'], age_bins, include_lowest=False, duplicates='drop')
new_purchase_data = purchase_data
new_purchase_data['Age Range'] = pd.cut(purchase_data['Age'], age_bins, labels=age_labels, include_lowest=False, duplicates='drop')

In [44]:
purchas_analysis_grouped = new_purchase_data.groupby(['Age Range'])
purchas_analysis_grouped = purchas_analysis_grouped.agg({
    'Purchase ID': ['count'],
    'Price': ['mean', 'sum'],
    'SN': 'nunique'
})
purchas_analysis_grouped.columns = purchas_analysis_grouped.columns.droplevel(0)
purchas_analysis_age_df = pd.DataFrame(purchas_analysis_grouped)
purchas_analysis_age_df = purchas_analysis_age_df.rename(columns={
    'count':'Purchase Count',
    'mean': 'Average Purchase Price',
    'sum': 'Total Purchase Value',
    'nunique': 'AVG Total per Person By Age'
})
purchas_analysis_age_df['AVG Total per Person By Age'] = purchas_analysis_age_df.apply(lambda row: row['Total Purchase Value']/row['AVG Total per Person By Age'], axis=1)
purchas_analysis_age_df['Average Purchase Price'] = purchas_analysis_age_df['Average Purchase Price'].map('${:.2f}'.format)
purchas_analysis_age_df['Total Purchase Value'] = purchas_analysis_age_df['Total Purchase Value'].map('${:.2f}'.format)
purchas_analysis_age_df['AVG Total per Person By Age'] = purchas_analysis_age_df['AVG Total per Person By Age'].map('${:.2f}'.format)
purchas_analysis_age_df.index.name = None
purchas_analysis_age_df

Purchase Count  ... AVG Total per Person By Age
<10                23  ...                       $4.54
10-14              28  ...                       $3.76
15-19             136  ...                       $3.86
20-24             365  ...                       $4.32
25-29             101  ...                       $3.81
30-34              73  ...                       $4.12
35-39              41  ...                       $4.76
40+                13  ...                       $3.19

[8 rows x 4 columns]

### Result Top Spenders

* Identify the the top 5 spenders in the game by total purchase value, then list (in a table):
  * SN
  * Purchase Count
  * Average Purchase Price
  * Total Purchase Value
 ----------

In [45]:
top_spenders_grouped = purchase_data.groupby('SN')
top_spenders_grouped = top_spenders_grouped.agg({
    'SN': 'count',
    'Price': ['mean', 'sum'],
})
top_spenders_grouped.columns = top_spenders_grouped.columns.droplevel(0)
top_spenders_grouped
top_spenders_df = pd.DataFrame(top_spenders_grouped)
top_spenders_df = top_spenders_df.sort_values('sum', ascending=False)
top_spenders_df = top_spenders_df.rename(columns={
    'count': 'Purchase Count',
    'mean': 'Average Purchase Price',
    'sum': 'Total Purchase Value'
})
top_spenders_df['Average Purchase Price'] = top_spenders_df['Average Purchase Price'].map('${:.2f}'.format)
top_spenders_df['Total Purchase Value'] = top_spenders_df['Total Purchase Value'].map('${:.2f}'.format)
top_spenders_df.index.name = None
top_spenders_df.head()

Purchase Count Average Purchase Price Total Purchase Value
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

### Result Most Popular Items

* Identify the 5 most popular items by purchase count, then list (in a table):
  * Item ID
  * Item Name
  * Purchase Count
  * Item Price
  * Total Purchase Value
 ---------

In [46]:
most_popular_item_grouped = purchase_data.groupby(['Item ID','Item Name'])
most_popular_item_grouped = most_popular_item_grouped.agg({
    'Item ID': 'count',
    'Price': 'sum'
})
most_popular_item_df = most_popular_item_grouped
most_popular_item_grouped = most_popular_item_grouped.sort_values('Price', ascending=False)
most_popular_item_df['Item Price'] = most_popular_item_df.apply(lambda row: row[1]/row[0], axis=1)
most_popular_item_df = most_popular_item_df.rename(columns={
    'Item ID': 'Purchase Count',
    'Price': 'Total Purchase Value'
}) 
most_profitable_item_df = most_popular_item_df
most_popular_item_df = most_popular_item_df[['Purchase Count', 'Item Price', 'Total Purchase Value']]
most_popular_item_df['Item Price'] = most_popular_item_df['Item Price'].map('${:.2f}'.format)
most_popular_item_df['Total Purchase Value'] = most_popular_item_df['Total Purchase Value'].map('${:.2f}'.format)
most_popular_item_df = most_popular_item_df.sort_values('Purchase Count', ascending=False)
most_popular_item_df.head()

Purchase Count  ... Total Purchase Value
Item ID Item Name                                                     ...                     
92      Final Critic                                              13  ...               $59.99
178     Oathbreaker, Last Hope of the Breaking Storm              12  ...               $50.76
145     Fiery Glass Crusader                                       9  ...               $41.22
132     Persuasion                                                 9  ...               $28.99
108     Extraction, Quickblade Of Trembling Hands                  9  ...               $31.77

[5 rows x 3 columns]

### Result Most Profitable Items

* Identify the 5 most profitable items by total purchase value, then list (in a table):
  * Item ID
  * Item Name
  * Purchase Count
  * Item Price
  * Total Purchase Value
 --------

In [47]:
most_profitable_item_df = most_profitable_item_df[['Purchase Count', 'Item Price', 'Total Purchase Value']]
most_profitable_item_df = most_profitable_item_df.sort_values('Total Purchase Value', ascending=False)
most_profitable_item_df['Item Price'] = most_profitable_item_df['Item Price'].map('${:.2f}'.format)
most_profitable_item_df['Total Purchase Value'] = most_profitable_item_df['Total Purchase Value'].map('${:.2f}'.format)
most_profitable_item_df.head()

Purchase Count  ... Total Purchase Value
Item ID Item Name                                                     ...                     
92      Final Critic                                              13  ...               $59.99
178     Oathbreaker, Last Hope of the Breaking Storm              12  ...               $50.76
82      Nirvana                                                    9  ...               $44.10
145     Fiery Glass Crusader                                       9  ...               $41.22
103     Singed Scalpel                                             8  ...               $34.80

[5 rows x 3 columns]

## Observable Trends

### We have a considerable male population
According to the results more than 80% of our players are males, this means that we should focus at least 20% of our efforts with campaigns, items and marketing to this population to satisfy their needs to continue scaling our game. We should remember the 80/20 pareto rule and understand that most of our revenue will come from them and this efforts must include our biggest changes.

### Ages 15 to 24
The data shows that our main player population has an age of 15 to 24 this could mean that they are more willing to spend since probably they still have economic support from their parents. We should consider a marketing campaign focus on highschool and university students with a back to school or student program to grow this market segment.

### Item price for male population
As we can see \\$4.47 is the average total purchase per male, so we can introduce items that range from \\$4.00 to \\$5.00 sepecific for the male gender, like aestehtics items with fire, spikes, and exaggerated names such as 'Critical Sword of Destiny' ,vanity items to change appearance such as 'Height, Muscle body, hairstyles and titles' and other types which focus on game experience ingame quests and more.
